### ISSUE TITLE
[Game Play] Rework the movement system

### ISSUE THREAD

**Issue by EvoLandEco** [tresinformal/game#524]
I feel like the current movement control is not smooth. There is a test of whether inertia slows down players, though not robust. IMO the movement system can to be optimized, and more robust tests should be added.

Despite of fixing the bugs, I would also like to make the movement system more flexible and player-friendly, the new system would be able to provide a smoother control of the protagonists, and to allow for further in-depth optimization.

I propose to give the `player` class the following parameters:

- `player_max_speed_forward` the maximum speed when moving forward
- `player_max_speed_backward` the maximum speed when moving backward
- `player_acceleration_forward` the acceleration forward when we press and hold the move forward key
- `player_acceleration_backward` the acceleration backward when we press and hold the move backward key
- `player_deceleration_forward` the deceleration due to inertia when moving forward, if we do not press the keys to move forward or backward
- `player_deceleration_backward` the deceleration due to inertia when moving backward, if we do not press the keys to move forward or backward

By doing this the players would be able accelerate and decelerate naturally and differently between forward and backward directions.

`action_type` class will be modified accordingly:

`action_type::brake` will be removed
`action_type::accelerate` will be changed to `action_type::accelerate_forward`
`action_type::acc_backward` will be changed to `action_type::accelerate_backward`

I also propose to give the `player` class a member variable of action flag, which indicates what the player is doing during the current tick, this enables us to do many things in the future, and at the moment, the ability to allow `game::apply_inertia()` to judge whether the player is pressing the keys to move forward or backward. If so, `game::apply_inertia()` should not work, and if not, the players should auto-decelerate due to the inertia.

I don't think `brake` is needed, firstly we already have inertia (with value large enough) which behaves as auto-brake, secondly this makes the control more difficult and less enjoyable, unless the effect of inertia is really small that we need to brake manually, and we want to make a relatively hard-core game.

Please advice me if there is a better way to implement this idea, and comment if you agree/disagree with me😀

**Comment by richelbilderbeek** [comment]
Looks solid to me!

Here are my thoughts. Feel free to ignore all of these if you want to make your own mistakes (and you are able to do so!) :-)

 * Do split this up in multiple Issue and refer these from this 'master Issue'. We want to follow the mantra 'commit early, commit often', i.e. you should not break the game too long and work with small iterations instead
 * Create a class called `player_options`/`player_params`/`player_settings` (pick you favorite name!) that contains all these player parameters. Do not create `player_options::player_[x]`, but call it `player_options::m_[x]`
 * I think `acceleration_forward` is a misnomer that is part in the current code; IIRC in physics the change in acceleration is called either the `relative_accelaration`/`delta_accelaration` and this is caused by an impulse (hence variabel name `m_impulse_forward`). 
 * There is a difference in braking and accelerating backwards: if you brake, you speed always decreases. If you accelarate backwards, you speed increases if you move backwards. There are some tests that protect that principle. Keep those. If you really think this movement makes the game duller, we'll vote, but I feel that would be way too early :-)

Keep up the good work!

**Comment by richelbilderbeek** [comment]
Ah, I see there is a Pull Request already :-) . If things are already done, I will accept, as I feel that going fast and breaking things occasionally is better than being afraid to break things. Or: I enjoy the recklessness/ruthlessness!

**Comment by EvoLandEco** [comment]
> Looks solid to me!
> 
> Here are my thoughts. Feel free to ignore all of these if you want to make your own mistakes (and you are able to do so!) :-)
> 
> * Do split this up in multiple Issue and refer these from this 'master Issue'. We want to follow the mantra 'commit early, commit often', i.e. you should not break the game too long and work with small iterations instead
> * Create a class called `player_options`/`player_params`/`player_settings` (pick you favorite name!) that contains all these player parameters. Do not create `player_options::player_[x]`, but call it `player_options::m_[x]`
> * I think `acceleration_forward` is a misnomer that is part in the current code; IIRC in physics the change in acceleration is called either the `relative_accelaration`/`delta_accelaration` and this is caused by an impulse (hence variabel name `m_impulse_forward`).
> * There is a difference in braking and accelerating backwards: if you brake, you speed always decreases. If you accelarate backwards, you speed increases if you move backwards. There are some tests that protect that principle. Keep those. If you really think this movement makes the game duller, we'll vote, but I feel that would be way too early :-)
> 
> Keep up the good work!

The impulse point of view is quite interesting, this makes me think of a racing game like Forza Horizon 4, the player moves like a car. In this case we set a very small inertia effect, then `brake` is important here just like a handbrake in a car. 

I was thinking of a game that the player moves like a human in an FPS game, or a heavy tank in the World of Tanks. In both cases `brake` is not important.

Anyway, I agree to add `brake` back in, which is inline with what I proposed: to make the movement system more flexible. I will break this comment into several issues and implement them one by one.

**Comment by EvoLandEco** [comment]
Let's make the implementation in PR #525  clearer:

- Although `brake` is moved, players do have one deceleration methods that considers both directions `decelerate()` , associated with two parameters determining how fast they decelerate in both directions.
- When a player is moving in one direction (e.g. forward), pressing the button to accelerate to the opposite direction (e.g. backward) will not trigger acceleration to the opposite direction (e.g. `accelerate_backward`). Rather, `decelerate()` will be triggered.
-  A player cannot 'over-decelerate', which means that `decelerate()` will only change the speed towards 0. When a player's speed reaches 0, then `accelerate_forward/backward` will take over.
- If you press nothing, then the inertia will also trigger `decelerate()`. So the inertia has exactly the same effect as pressing the button to accelerate to the opposite direction.
- To summarize, in PR #525, `brake` is automatic, if you press the button to accelerate to the opposite direction against the direction you are moving to.

To put `brake` back in, we may need to:

- Isolate inertia from `decelerate()`, so that inertia has its own parameter, probably player-specific: `m_player_inertia`.
- Rename `brake` to `handbrake`;
- Add a player-specific parameter `m_player_handbrake_speed`, along with a method `player::handbrake()`.
- Add `action_type::handbrake`, find a new key and map it to the action.
- Just an idea: the effect of handbrake should be much stronger than auto-brake or the inertia.

**Comment by EvoLandEco** [comment]
closed by #524
